source/plugins/ruby/in_kube_podinventory.rb

```diff
@@ -535,25 +394,238 @@ def run_periodic
+
+              #Populate mdm metric for container restart count if greater than 0
+              if (!containerRestartCount.nil? && (containerRestartCount.is_a? Integer) && containerRestartCount > 0)
+                @inventoryToMdmConvertor.process_record_for_container_restarts_metric(record["ControllerName"], record["Namespace"], lastFinishedTime)
+              end
+            rescue => errorStr
+              $log.warn "Failed in parse_and_emit_record pod inventory while processing ContainerLastStatus: #{errorStr}"
+              $log.debug_backtrace(errorStr.backtrace)
+              ApplicationInsightsUtility.sendExceptionTelemetry(errorStr)
+              record["ContainerLastStatus"] = Hash.new
+            end
+
+            podRestartCount += containerRestartCount
+            records.push(record.dup)
+          end
+        else # for unscheduled pods there are no status.containerStatuses, in this case we still want the pod
+          records.push(record)
+        end  #container status block end
+
+        records.each do |record|
+          if !record.nil?
+            record["PodRestartCount"] = podRestartCount
+          end
+        end
+      rescue => error
+        $log.warn("getPodInventoryRecords failed: #{error}")
+      end
+      return records
+    end
+
+    # TODO - move this method to KubernetesClient or helper class
+    def getServiceNameFromLabels(namespace, labels, serviceRecords)
       serviceName = ""
       begin
         if !labels.nil? && !labels.empty?
-          if (!serviceList.nil? && !serviceList.empty? && serviceList.key?("items") && !serviceList["items"].empty?)
-            serviceList["items"].each do |item|
-              found = 0
-              if !item["spec"].nil? && !item["spec"]["selector"].nil? && item["metadata"]["namespace"] == namespace
-                selectorLabels = item["spec"]["selector"]
-                if !selectorLabels.empty?
-                  selectorLabels.each do |key, value|
-                    if !(labels.select { |k, v| k == key && v == value }.length > 0)
-                      break
-                    end
-                    found = found + 1
+          serviceRecords.each do |kubeServiceRecord|
+            found = 0
+            if kubeServiceRecord["Namespace"] == namespace
+              selectorLabels = {}
+              # selector labels wrapped in array in kube service records so unwrapping here
+              if !kubeServiceRecord["SelectorLabels"].nil? && kubeServiceRecord["SelectorLabels"].length > 0
+                selectorLabels = kubeServiceRecord["SelectorLabels"][0]
+              end
+              if !selectorLabels.nil? && !selectorLabels.empty?
+                selectorLabels.each do |key, value|
+                  if !(labels.select { |k, v| k == key && v == value }.length > 0)
+                    break
                   end
+                  found = found + 1
                 end
+                # service can have no selectors
```

Comment:
can you explain this change ? We discover only services based on label selectors.

Reply:
as discussed in the email, this is just not  tagging service name if the service has zero selectors and service and pod are in the same namespace.  saw one of Prometheus service doesnt have selectors. since initial value of found =0, found == selectorLabels.length. This will be true if the length of service labal selector is 0.